compiler/src/dotty/tools/dotc/typer/Namer.scala

```diff
@@ -1556,6 +1556,17 @@ class Namer { typer: Typer =>
     val termParamss = ctx.normalizeIfConstructor(vparamss.nestedMap(symbolOfTree), isConstructor)
     sym.setParamss(typeParams, termParamss)
     def wrapMethType(restpe: Type): Type = {
+      restpe match
+        case tv: TypeVar
```

Comment:
Checking for type variables at the top-level is not enough, for example I can reproduce the ClassCastException by wrapping the result type in an identity type lambda:
```scala
object Test {
  type Id[T] = T

  sealed trait Container { s =>
    type A
    def visit[R](int: IntV & s.type => Id[R], str: StrV & s.type => Id[R]): R
  }
  final class IntV extends Container { s =>
    type A = Int
    val i: Int = 42
    def visit[R](int: IntV & s.type => Id[R], str: StrV & s.type => Id[R]): R = int(this)
  }
  final class StrV extends Container { s =>
    type A = String
    val t: String = "hello"
    def visit[R](int: IntV & s.type => Id[R], str: StrV & s.type => Id[R]): R = str(this)
  }

  def minimalOk[R](c: Container { type A = R }): R = c.visit[R](
    int = vi => vi.i : Id[vi.A],
    str = vs => vs.t : Id[vs.A]
  )
  def minimalFail[M](c: Container { type A = M }): M = c.visit(
    int = vi => vi.i : Id[vi.A],
    str = vs => vs.t : Id[vs.A]  // error
  )

  def main(args: Array[String]): Unit = {
    val e: Container { type A = String } = new StrV
    println(minimalOk(e)) // this one prints "hello"
    println(minimalFail(e)) // this one fails with ClassCastException: class java.lang.String cannot be cast to class java.lang.Integer
  }
}
```
(more complex types are also probably problematic but slightly harder to get a ClassCastException out of)
More complex types